### PR TITLE
RavenDB-27286 Add pattern matching and wildcard support in Corax queries

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-27286

### Additional description

Support for queries like:
`where search(Field, "m*ej")`
`where search(Field, "?aciej")`

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

**Comment by arekpalinski** [comment]
@maciejaszyk please check the conflict

**Comment by arekpalinski** [comment]
@ravendb-bot review this please

**Comment by redknightlois** [comment]
_Automated review (hugin). Posted manually — the bot's own post step was blocked by a false positive in its output-safety check._

---

## Overview

This PR adds glob-style pattern matching (`*` and `?` wildcards) to Corax search queries. It introduces a `PatternMatch` search option, a `PatternTermProvider` iterator that walks the compact tree with prefix-seeking and matches remaining terms against the pattern, and wires the new provider into the multi-term-match dispatch chain in `IndexSearcher`.

## Changes

- `src/Corax/Constants.cs` — Adds `Asterisk`, `QuestionMark`, `PatternSymbols` search values, and the `PatternMatch` flag to `SearchMatchOptions`.
- `src/Corax/Querying/IndexSearcher.Search.cs` — Rewrites `GetTermType` to detect patterns containing `*` or `?`; adds `PatternMatch` to the search-query routing switch.
- `src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs` (new) — `PatternTermProvider<TLookupIterator>` struct: seeks the compact tree forward to the literal prefix before the first wildcard, then walks remaining terms applying an ASCII/UTF-8 glob matcher.
- `src/Corax/Querying/MultiTermMatches/` — Adds `PatternQuery` factory method and the forward/backward provider dispatch cases in `GetMultiTermMatchProvider`.
- Tests — `test/FastTests/Corax/IndexSearcher.cs` exercises `PatternQuery` directly with 16 pattern cases; `test/SlowTests/Issues/RavenDB_27286.cs` validates end-to-end `Search()` behaviour across both Corax and Lucene engines.

<details><summary>Show a summary per file</summary>

| File | Description |
|------|-------------|
| `src/Corax/Constants.cs` | Pattern symbols, `SearchValues<byte>`, and `PatternMatch` enum flag |
| `src/Corax/Querying/IndexSearcher.Search.cs` | `GetTermType` rewrite and `PatternMatch` dispatch |
| `src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs` | New `PatternTermProvider<TLookupIterator>` glob-matching iterator |
| `src/Corax/Querying/MultiTermMatches/IndexSearcher.MultiTermMatch.Simple.cs` | `PatternQuery` factory method |
| `src/Corax/Querying/MultiTermMatches/IndexSearcher.MultiTermMatches.Common.cs` | Forward/backward provider dispatch cases |
| `test/FastTests/Corax/IndexSearcher.cs` | 16-case direct `PatternQuery` test |
| `test/SlowTests/Issues/RavenDB_27286.cs` | Cross-engine integration test via `Search()` |

</details>

## Tests

- The fast test (`WildcardQueryStatement`) exercises `PatternQuery` directly with 16 pattern/expected-count pairs including edge cases like consecutive `**`, leading/trailing `*`, `?` mixing, and Unicode. Correctly single-mode — it uses `IndexSearcher.PatternQuery()` directly, which is an engine-specific API call.
- The slow test (`SearchWithWildcardInsideTermBehavesTheSameOnBothEngines`) parameterises 8 pattern cases across `RavenSearchEngineMode.All`, verifying Corax and Lucene produce identical result counts through the public `Search()` path. Correct categorisation (`RavenTheory`, `RavenTestCategory.Querying`).

## Coverage

### Corax

| Dimension | Status |
|-----------|--------|
| Concurrency — Searcher snapshot publication | N/A — no per-field cache or side structure added |
| Concurrency — Searcher iteration contract | Covered — `PatternTermProvider.Next` yields between `MoveNext` calls holding iterator and buffer state |
| Concurrency — Cross-engine boundaries | N/A — no Corax/Lucene bridge code changed |
| Lifetime — IndexSearcher lifetime | Covered — struct captures `_searcher` reference valid for query scope only |
| Lifetime — Per-query buffers | Covered — `_termBuffer` allocated from searcher allocator, struct is short-lived |
| Lifetime — Per-field caches on persistence | N/A — no persistence-layer cache added |
| Lifetime — Native handle in value type | N/A — no native handles in the new struct |
| On-disk compatibility | N/A — all changes are in-memory (search option flag, pattern symbols, iterator); no format, encoding, or metadata change |
| Performance — PERF-marked match operators | N/A — no `PERF` annotations on changed methods |
| Performance — Hot-path allocations | Covered — one `_termBuffer` ByteString allocation per `PatternTermProvider` construction, scoped to the allocator |
| Performance — Pool reuse | N/A — no existing term-provider pool to wire into; term providers are inherently per-query |
| Performance — Vectorised match operators | N/A — no SIMD code changed |

## Review

| Severity | Count |
|----------|-------|
| 🔴 Bug   | 1 |
| 🟡 Concern | 0 |

## Finding

`src/Corax/Querying/IndexSearcher.Search.cs:514`

🔴 **Bug — GetTermType does not route ?-only or interior-* patterns to PatternMatch**

The byte `GetTermType` returns `TermMatch` for patterns like `?aciej` and `m*ej` — examples the PR description claims to support — because the `PatternMatch` guard requires a leading `*` (`hasPrefixAsterisk && hasLeftTermAnyPatterns`), and patterns whose first byte is not `*` fall through to the `TermMatch` / `StartsWith` / `EndsWith` fallbacks. The `PatternMatch` dispatch arm exists downstream in the switch and `PatternQuery` / `PatternTermProvider` work correctly (the fast test proves this), but the search API routing never reaches them for these inputs, silently treating the wildcard characters as literals.

**Proof:** Walkthrough for `search(Field, "?aciej")` through `SearchQueryWithPhraseQueryWithWildcardQueriesAdjustments` (the default `SearchQuery` path):

1. The string `"?aciej"` is tokenised as a single term, encoded to UTF-8 bytes, and at the line that reads `termType = GetTermType(valueAsSpan)`, the byte overload of `GetTermType` is called with the six UTF‑8 bytes of `?aciej`.
2. `termValue.ContainsAny(Constants.Search.PatternSymbols)` is `true` (the term contains the byte `'?'`) → does *not* short-circuit to `TermMatch`.
3. `hasPrefixAsterisk = termValue[0] == '*'` → `false` (first byte is `'?'`).
4. `hasSuffixAsterisk = termValue[^1] == '*' && (…)` → `false` (last byte is `'j'`).
5. The all‑asterisks `Exists` check (`hasPrefixAsterisk & hasSuffixAsterisk && …`) falls through.
6. `leftTerm = termValue[0..6]` = the full `"?aciej"`. `hasLeftTermAnyPatterns` = `true`.
7. **The only gate to `PatternMatch`:** `hasPrefixAsterisk && hasLeftTermAnyPatterns` → `false && true` → **`false`**.
8. None of the subsequent conditions match; the function reaches `hasSuffixAsterisk ? StartsWith : TermMatch` and returns **`TermMatch`**.
9. Back in the caller, `termType is TermMatch` is `true` → the term is added to `termMatches` and later searched as a literal byte sequence `?aciej`. The `PatternMatch` arm of the downstream switch and `PatternQuery` are never reached.

The same trace produces `TermMatch` for `m*ej`, `a?b`, and any pattern whose first byte is not `*` but that contains `?` or an interior `*`. The fix is to return `PatternMatch` when `hasLeftTermAnyPatterns` is `true`, placed after the `Exists` short‑circuit and before the `Contains` / `EndsWith` / `StartsWith` / `TermMatch` fallbacks.

## Verdict

1 blocking issue to address before merge — `GetTermType` silently treats `?-only` and `interior-*` patterns as literal terms instead of routing them to `PatternMatch`.

File: src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+    public bool IsFillSupported => false;
+
+    public int Fill(Span<long> containers) => throw new NotImplementedException();
+
+    public void Reset()
+    {
+        _iterator = _tree.Iterate<TLookupIterator>();
+
+        // Only the forward iterator can seek to the prefix; the backward iterator falls back to a full scan.
```
Comment (ayende) [inline src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs]: we can seek _after_ the current prefix, then scan backward, no?

File: src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        _field = field;
+        _pattern = pattern;
+        _token = token;
+        _tree = tree;
+
+        var patternSpan = pattern.Decoded();
+
+        // The literal run before the first wildcard character lets us seek instead of scanning the whole field.
+        var firstPatternCharacter = patternSpan.IndexOfAny(Constants.Search.PatternSymbols);
+        _seekPrefixLength = firstPatternCharacter == -1 ? patternSpan.Length : firstPatternCharacter;
```
Comment (ayende) [inline src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs]: If we don't find a pattern, isn't that  an exact match?

File: src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        else
+            _iterator.Reset();
+    }
+
+    public bool Next(out TermMatch term)
+    {
+        var pattern = _pattern.Decoded();
+        var prefix = default(TLookupIterator).IsForward ? pattern[.._seekPrefixLength] : default;
+
+        while (_iterator.MoveNext(out var compactKey, out _, out _))
```
Comment (ayende) [inline src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs]: here you allocate a new compact key per iteration? Or reuse a single one?
That was a common cause for perf issues in Corax 2.0, and was fixed in a few places in 7.2

File: src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs (hunk 2 above)
Comment (maciejaszyk) [reply src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs]: Actually it is, and it should not be used as Wildcard in such case, because then it's a TermMatch. Added an exception.

File: src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs (hunk 3 above)
Comment (maciejaszyk) [reply src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs]: Yes, fixed to allocate only once.

File: src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Corax/Querying/Matches/TermProviders/TermProvider.Wildcard.cs]: Done

**Comment by maciejaszyk** [comment]
Regarding bug found in @redknightlois - this way we're compatible between search engines, if someone wants to find term with "?" or "*" it should use equals.

**Review by ayende — APPROVED** [review]

**Review by arekpalinski — APPROVED** [review]
